libyul/ObjectParser.cpp

```diff
@@ -104,9 +110,80 @@ shared_ptr<Block> ObjectParser::parseCode()
 	return parseBlock();
 }
 
+optional<ObjectParser::ReverseSourceNameMap> ObjectParser::tryGetSourceLocationMapping() const
+{
+	return tryGetSourceLocationMapping(
+		m_scanner->currentCommentLiteral(),
+		m_scanner->currentCommentLocation(),
+		m_errorReporter
+	);
+}
+
+optional<ObjectParser::ReverseSourceNameMap> ObjectParser::tryGetSourceLocationMapping(
+	std::string const& _text,
+	SourceLocation const& _location,
+	ErrorReporter& _errorReporter
+)
+{
+	// @use-src 0:"abc.sol", 1:"foo.sol", 2:"bar.sol"
+	//
+	// UseSrcList := UseSrc (',' UseSrc)*
+	// UseSrc     := [0-9]+ ':' FileName
+	// FileName   := "(([^\"]|\.)*)"
+
+	// Matches some "@use-src TEXT".
+	static std::regex const lineRE = std::regex(
+		R"~~~((^|\s+)@use-src\s+(.*)$)~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+	std::smatch sm;
+	if (!std::regex_search(_text, sm, lineRE))
+		return nullopt;
+	solAssert(sm.size() == 3, "");
+
+	// Let @c text point to the parameter value (last match).
+	auto text = sm[2].str();
+
+	// iteratively match for NUM : STRING_LITERAL and increment
+	static regex const firstParamRE(R"~~~(\s*(\d+)\s*:\s*"((?:\\\"|[^\"])*)")~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+	static regex const continuationParamRE(R"~~~(\s*,\s*(\d+)\s*:\s*"((?:\\\"|[^\"])*)")~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+
+	ReverseSourceNameMap result;
+
+	while (!text.empty())
+	{
+		if (!std::regex_search(text, sm, !result.empty() ? continuationParamRE : firstParamRE))
+		{
+			_errorReporter.syntaxError(9804_error, _location, "Could not fully parse @use-src arguments.");
+			return nullopt;
+		}
+		solAssert(sm.size() == 3, "");
+
+		auto const len = sm[0].length();
+		solAssert(len > 0, "");
+
+		auto const sourceIndex = toUnsignedInt(sm[1].str());
+		if (!sourceIndex)
+		{
+			_errorReporter.syntaxError(1619_error, _location, "Invalid parameters for @use-src."); // TODO: error code
+			return nullopt;
+		}
+
+		auto sourceName = make_shared<string const>(sm[2].str());
```

Comment:
is this local variable useful?

Reply:
If you use it to make it clear that it is a source name, why not rename `result` to `sourceNames`?

Reply:
Oh and you have to unescape the file name.